sql/opt_range.cc

```diff
@@ -1636,6 +1636,26 @@ int QUICK_ROR_INTERSECT_SELECT::init()
 }
 
 
+/*
+  Mark column bitmap for this index used by ROR-merged scan
+*/
+void QUICK_RANGE_SELECT::mark_columns_ror_merged_scan() {
+  /*
+    We are only going to read key fields and call position() on 'file'
+    The following sets head->tmp_set to only use this key and then updates
+    head->read_set and head->write_set to use this bitmap.
+    The now bitmap is stored in 'column_bitmap' which is used in ::get_next()
+  */
+  handler *org_file= head->file;
+  head->file= file;
+  /* We don't have to set 'head->keyread' here as the 'file' is unique */
+  if (!head->no_keyread)
+    head->mark_columns_used_by_index(index);
+  head->prepare_for_position();
```

Comment:
Do we really get here when head->no_keyread==true? (As far I remember, this happens when the table is updated, and index_merge is not used in such cases.  If one is not allowed to use index-only reads, then there is no point to use index_merge, because you're going to read the same full row using multiple indexes...)

Reply:
This is old code that I've refactored into this function, so I'm not sure exactly when it is required. Looking at the history, this [commit](https://github.com/mysql/mysql-server/commit/28d3ee656df4d97bee8b3cf1554bef3120cdc8a4) added it in 2006 but unfortunately had little information. Interestingly, I did find [this MariaDB bug report](https://bugs.launchpad.net/maria/+bug/1006164) that seems to indicate that this scenario could be triggered by `DELETE tbl FROM tbl WHERE key1=const1 AND key2=const2;` but the scenario in question doesn't seem to actually work and we might need to call `mark_columns_used_by_index` regardless, based on your investigation in the bug back in 2012. If we want to change it I think it should be a separate diff with a test case similar to what's in mariadb bug.